### ISSUE TITLE
import as

### ISSUE THREAD

**Issue by idavydov** [rticulate/import#27]
One of the frequent patterns in python looks like this:
```python
import pandas as pd
import numpy as np
```

Would it be possible to have something like this in the `import` package?

Ideally, something like this: 
```r
import::from("magrittr", "%>%")
import::as(dp="dplyr", tr="tidyr", pr="purrr")

cars %>%
  dp::select(...) %>%
  tr::pivot_wider(...)
```

In theory this might be possible in case `::` operator is extended.

Or maybe there's an easier way to do this?

**Comment by smbache** [comment]
```R
s <- import::from(stats, acf)

s$acf
```

**Comment by idavydov** [comment]
Thanks, @smbache! Nice. But this still:
a) populates the namespace
b) does not allow to import all function

Is there something similar to this:

```
dp <- import::from('dplyr', '*', .detacth=TRUE)
```

Now it might be possible with something like this:
```
# this does not work, you cannot pass a character vector
dp <- import::from('dplyr', getNamespaceExports('dplyr'), .into='imports:dplyr')
detach('imports:dplyr')
```

**Comment by smbache** [comment]
re b) this is a part of [this pr](https://github.com/smbache/import/pull/26),

re a) I guess it's possible to do more what you want, although i'm unsure whether it is generally desirable..  it wouldn't be via the `::` though. but the environment (which is returned in the above example) itself could be exposed in `imports`. Whether it's worth it 🤷‍♂️

**Comment by idavydov** [comment]
Thanks, @smbache !

Yep, whether `$` or `::` is not important. So it would be something like this, then?

```r
import::from(magrittr, '%>%')
dp <- import::from(dplyr, .all=TRUE, .into=new.env())

cars %>%
  dp$select(...)
```

Whether desirable, in my opinion short names is a good compromise between importing everything in the namespace and typing `dplyr::` each time. But of course some would disagree.

**Comment by smbache** [comment]
But what's the win over having it separately namespaced? It won't clutter you global namespace.

**Comment by idavydov** [comment]
@smbache, not sure I understand what do you mean by separately namespaces?

I think your first example pollutes the global namespace. An example would be extremely helpful.

**Comment by torfason** [comment]
Judging from my reading of the code , it would have to be:

`dp <- import::from(dplyr, .all=TRUE, .into={new.env()})`

But even then it would not work, because the last line of `from()` reads:

`invisible(if (use_into) as.environment(into) else NULL)`

And in the above case, the test causes NULL to be returned. That said, I'm not sure why it is important to do that test, and have not found any use cases in my tests that cause errors if the line is changed to simply:

`invisible(as.environment(into))`

@smbache, I wonder if I should make this change in my pull request?  Is the test simply intended to avoid returning the global environment? Is there any issue with doing that?

I have to admit that I am struggling a bit with the `.into` parts of the code, and don't recall seeing any documentation of the `{env}` notation.  It also seems to me that using `.into=""` fails silently using the released version. Wouldn't it be cleaner to require the use of `here()` to import into the current environment? I'll try to address these issues in my updated pull request and include tests for as many use cases as possible, so that the decision on desired behavior is explicitly shown.

**Comment by smbache** [comment]
@torfason I think it's fine to return invisbly and omit the check. docs would need to be updated.

the `{env}` notation is to make the environment be treated by value, rather than inferred by name. Interesting that I never documented that 🤷‍♂️ 

The `.into=""` should not fail, and we should look `.into` that :)

**Comment by torfason** [comment]
`.into=""` should now be fixed in the pull request.

**Comment by torfason** [comment]
This should now be fixed on master. If you could install the most current version and test whether the fix works for you (and if the new version works for you in general), that would be greatly appreciated. You can get the fix with:
`remotes::install_github("smbache/import")`

The suggested approach is to use `.into={new.env()}` to assign to a new environment, and assign the returned environment to a variable:

``` r
# Import into a new namespace, use $ to access
dp <- import::from(dplyr, .all=TRUE, .into={new.env()})
dp$select(head(cars),dist)
#>   dist
#> 1    2
#> 2   10
#> 3    4
#> 4   22
#> 5   16
#> 6   10

# Functions are not visible without dp$ prefix
select(head(cars),dist)
#> Error in select(head(cars), dist): could not find function "select"
```






[ fixed-in-master : Issue to be closed following CRAN release ]

**Comment by idavydov** [comment]
Thanks. Works like a charm, @torfason !

**Comment by torfason** [comment]
Fixed in release 1.2.0
